crates/bevy_ecs/src/entity/map_entities.rs

```diff
@@ -56,20 +36,92 @@ pub trait MapEntities {
 ///
-/// This is typically used to coordinate data transfer between sets of entities, such as between a scene and the world or over the network.
-/// This is required as [`Entity`] identifiers are opaque; you cannot and do not want to reuse identifiers directly.
+/// This is typically used to coordinate data transfer between sets of entities, such as between a scene and the world
+/// or over the network. This is required as [`Entity`] identifiers are opaque; you cannot and do not want to reuse
+/// identifiers directly.
+///
+/// On its own, an `EntityMap` is not capable of allocating new entity identifiers, which is needed to map references
+/// to entities that lie outside the source entity set. To do this, an `EntityMap` can be wrapped in an
+/// [`EntityMapper`] which scopes it to a particular destination [`World`] and allows new identifiers to be allocated.
+/// This functionality can be accessed through [`Self::world_scope()`].
 #[derive(Default, Debug)]
 pub struct EntityMap {
     map: HashMap<Entity, Entity>,
 }
 
+/// A wrapper for [`EntityMap`], augmenting it with the ability to allocate new [`Entity`] references in a destination
+/// world. These newly allocated references are guaranteed to never point to any living entity in that world.
+///
+/// References are allocated by returning increasing generations starting from an internally initialized base
+/// [`Entity`]. After it is finished being used by [`MapEntities`] implementations, this entity is despawned and the
+/// requisite number of generations reserved.
+pub struct EntityMapper<'m> {
+    /// The wrapped [`EntityMap`].
+    map: &'m mut EntityMap,
+    /// A base [`Entity`] used to allocate new references.
+    dead_start: Entity,
+    /// The number of generations this mapper has allocated thus far.
+    generations: u32,
+}
+
+impl<'m> EntityMapper<'m> {
+    /// Returns the corresponding mapped entity or allocates a new dead entity if it is absent.
+    pub fn get_or_alloc(&mut self, entity: Entity) -> Entity {
+        if let Some(mapped) = self.map.get(entity) {
+            return mapped;
+        }
+
+        let new = Entity {
+            generation: self.dead_start.generation + self.generations,
+            index: self.dead_start.index,
+        };
+        self.generations += 1;
+
+        self.map.insert(entity, new);
+
+        new
+    }
+
+    /// Gets a reference to the underlying [`EntityMap`].
+    pub fn get_map(&'m self) -> &'m EntityMap {
+        self.map
+    }
+
+    /// Gets a mutable reference to the underlying [`EntityMap`]
+    pub fn get_map_mut(&'m mut self) -> &'m mut EntityMap {
+        self.map
+    }
+
+    /// Creates a new [`EntityMapper`], spawning a temporary base [`Entity`] in the provided [`World`]
+    fn new(map: &'m mut EntityMap, world: &mut World) -> Self {
+        Self {
+            map,
+            dead_start: world.spawn_empty().id(),
```

Comment:
Spawning a new entity here is a little hacky. Can we get a new index from `Commands`?

Reply:
Since getting rid of `try_reserve_generations` meant coupling this directly with `Entities` anyway, I went with getting a new index via `Entities::alloc()`.